### PR TITLE
[ML] Bootstrap estimate sample error in the test loss for training classification and regression models

### PR DESCRIPTION
**Pull request by tveasey**
We can use bootstrap estimates of the sample error in the test loss to better decide if we should continue searching hyperparameters: any variation which is small on the scale of the sample error is not worth burning runtime on.

Also, I've allowed us to decide just to use a hold out set if training data are plentiful rather than cross-validating. This will save us 50% of fine tune on large data sets.

This logically belongs with #2332: speed ups for large data sets. As such I don't want to document separately and marking as a non-issue.

### CODE REVIEW

File: lib/maths/common/CSampling.cc [hunk 1]
```diff
@@ -9,6 +9,7 @@
  * limitation.
  */
 
+#include <boost/random/poisson_distribution.hpp>
```
Comment (valeriy42) [inline lib/maths/common/CSampling.cc:12]: Nit: should go into the group below.

**Review by valeriy42 — APPROVED** [review]
LGTM. Good work. Looking forward to see performance results.

File: lib/maths/analytics/CBoostedTreeHyperparameters.cc [hunk 2]
```diff
@@ -480,13 +483,6 @@ bool CBoostedTreeHyperparameters::selectNext(const TMeanVarAccumulator& testLoss
         std::copy(m_HyperparameterSamples[m_CurrentRound].begin(),
                   m_HyperparameterSamples[m_CurrentRound].end(), parameters.data());
         parameters = minBoundary + parameters.cwiseProduct(maxBoundary - minBoundary);
-        if (m_CurrentRound == 0) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/analytics/CBoostedTreeHyperparameters.cc:483]: Yes, because we now have estimates of the variance in the loss values for all the earlier rounds. We could look at early exiting during fine tune on any iteration. I was going to explore this in separate change.
